src/queue.rs

```diff
@@ -0,0 +1,148 @@
+        match result {
+            Ok(Some(job)) => {
+                active_model.status = Set(JobStatus::Success);
+                active_model.scheduled_at = Set(None);
+                let mut next_job = ActiveModel::from(job);
+                next_job.parent_id = Set(Some(*active_model.id.as_ref()));
+                let next_job = next_job.insert(&tx).await?;
+                active_model.result = Set(Some(JobResult::Child(next_job.id)));
+            }
+
+            Ok(None) => {
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Success);
+                active_model.result = Set(Some(JobResult::Complete));
+            }
+
+            Err(e) if e.is_retryable() => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                let reschedule =
+                    schedule_based_on_failure_count(*active_model.failure_count.as_ref());
+                active_model.status =
+                    Set(reschedule.map_or(JobStatus::Failed, |_| JobStatus::Pending));
+                active_model.scheduled_at = Set(reschedule);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+
+            Err(e) => {
+                active_model.failure_count = Set(active_model.failure_count.as_ref() + 1);
+                active_model.scheduled_at = Set(None);
+                active_model.status = Set(JobStatus::Failed);
+                active_model.result = Set(Some(JobResult::Error(e)));
+            }
+        }
+        Some(active_model.update(&tx).await?)
+    } else {
+        None
+    };
+    tx.commit().await?;
+    Ok(model)
+}
+
+fn spawn(join_set: &mut JoinSet<()>, db: &Db, job_state: &Arc<SharedJobState>) {
+    let db = db.clone();
+    let job_state = Arc::clone(job_state);
+    join_set.spawn(async move {
+        loop {
+            match one(&db, &job_state).await {
+                Err(e) => {
+                    eprintln!("job error {e}");
+                }
+
+                Ok(Some(_)) => {}
+
+                Ok(None) => {
+                    sleep(Duration::from_secs(fastrand::u64(QUEUE_CHECK_INTERVAL))).await;
+                }
+            }
+        }
+    });
+}
+
+pub async fn run(db: Db, config: ApiConfig) {
+    let mut join_set = JoinSet::new();
+    let job_state = Arc::new(SharedJobState::from(&config));
+    for _ in 0..QUEUE_WORKER_COUNT {
+        spawn(&mut join_set, &db, &job_state);
+    }
+
+    while join_set.join_next().await.is_some() {
+        eprintln!("Worker task shut down. Restarting.");
+        spawn(&mut join_set, &db, &job_state);
+    }
+}
+
+async fn next(tx: &DatabaseTransaction) -> Result<Option<Model>, DbErr> {
+    let select = match tx.get_database_backend() {
+        backend @ DatabaseBackend::Postgres => Statement::from_sql_and_values(
+            backend,
+            r#"SELECT * FROM queue
```

Comment:
I think enumerating the fields you're `SELECT`ing is safer in the face of future schema migrations. If a column gets added to table `queue`, then this will start throwing an error (or crashing, or something bad) because the code that deserializes rows may not know how to deal with the new column.

Reply:
resolved by checking with the sea-orm folks and discovering that there is a somewhat-hidden way to add a locking clause to a query, so we can auto-enumerate the columns